### PR TITLE
"In-memory" repos

### PR DESCRIPTION
**Pull request by ethomson**
Allow fully in-memory repositories that don't point to a workdir or an on-disk git repository.  (At the moment, LibGit2Sharp users can hook up a custom object database backend, though not a refdb backend or custom configuration.)

### CODE REVIEW

File: LibGit2Sharp.Tests/RepositoryFixture.cs [hunk 1]
```diff
@@ -651,5 +651,20 @@ public void CanDetectShallowness()
...
+        {
+            using (var repo = new Repository())
+            {
+                /* TODO: bug in libgit2, update when fixed by
+                 * https://github.com/libgit2/libgit2/pull/2970
+                 */
+                // Assert.True(repo.Info.IsBare);
+
+                Assert.Equal(null, repo.Info.Path);
+                Assert.Equal(null, repo.Info.WorkingDirectory);
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/RepositoryFixture.cs]: Could we also assert that the `Index` is null?

**Comment by nulltoken** [comment]
>Allow fully in-memory repositories that don't point to a workdir or an on-disk git repository.

I like this trend!

While we're on this topic, one thing that really bothers me is that *currently* we plug the `OdbBackend`s through a setter. I'd really like those to be `xxBackend`moved to `RepositoryOptions`

File: LibGit2Sharp.Tests/RepositoryFixture.cs (hunk 1 above)
Comment (ethomson) [inline LibGit2Sharp.Tests/RepositoryFixture.cs]: Sure.  Interestingly, `git_repository_index()` doesn't return `null` here.  I was going to say well that's clever that it returns an in-memory index that isn't backed by a file!  But looking at the code I'm actually not sure what's going on now.  (https://github.com/libgit2/libgit2/blob/master/src/repository.c#L757)

So I think there might be an issue there in libgit2 that I need to dig into, but that's unrelated.  So I think that we should explicitly not try to load the index for in-memory repositories.  I'll push up a commit to that effect.

**Comment by ethomson** [comment]
You just want me to move `ObjectDatabase.AddBackend(OdbBackend backend, int priority)` -> `RepositoryInformation.AddObjectDatabaseBackend(OdbBackend backend, int priority)`?

**Comment by ethomson** [comment]
Oh, `RepositoryOptions`...  I think there's a chicken and egg problem there.  I don't think that I can necessarily create my `OdbBackend` before I create my repository...?

I mean, *I* can, especially in the in-memory case.  But I'm not sure that's generally so trivial...

File: LibGit2Sharp.Tests/RepositoryFixture.cs (hunk 1 above)
Comment (carlosmn) [inline LibGit2Sharp.Tests/RepositoryFixture.cs]: When you ask for an index at a particular path, we just  assume you don't mind it's empty because you'll be writing to it, a bit like the equivalent of

    GIT_INDEX=/tmp/altindex git read-tree

from git

Comment (ethomson) [inline LibGit2Sharp.Tests/RepositoryFixture.cs]: Yeah, but https://github.com/libgit2/libgit2/blob/master/src/repository.c#L757 assumes `path_repository` is non-NULL, which is not the case for repositories created with `git_repository_new`...

Comment (carlosmn) [inline LibGit2Sharp.Tests/RepositoryFixture.cs]: Oh right, sorry, of course we're missing the path completely. `git_buf_join{,path}()` seems to deal well with the first argument being NULL, which means we just use "index" as the path and would create it in the CWD, which is suboptimal.

So there should probably be a check before that join to see if we even have something to join and either return NULL or an error.

**Comment by ethomson** [comment]
So I didn't realize that the `Index` is never exposed on bare repositories in LibGit2Sharp.  I added a test to codify that behavior.  Because of that, we pick up those semantics for in-memory repositories as well.

I did *not* yet move the `OdbBackend`s to `RepositoryOptions`.  I think that should probably have a little bit more discussion.  Again, that's fine for my use case(s), but I worry that somebody might want to have a repository before they set up their custom object database backend.

For instance: you are some hypothetical program that uses the repository's configuration to identify some database that it uses as a backend.  Presently, we can open the repository, read the config, create the ODB backend and then set it on the repository.

If we must supply backends to the `RepositoryOptions` when we create the repository then we have to find some other mechanism to read the configuration...  Presumably nobody wants to re-create the config reading code, so this probably means something like:  open the repository, read the config, close the repository, set up the ODB backend, then reopen the repository (with the custom ODB backend).  This seems suboptimal.

**Comment by ethomson** [comment]
Further thoughts, @nulltoken ?

**Comment by nulltoken** [comment]
We're talking about allowing a repository to potentially not rely on the file system at all (eg. in-memory storage, database, ..). #900 by @zoltu deals specifically with this kind of request.

To me, relying on the default config file to hold some of the required parameters to set up the backends defeats a bit this purpose. Going a bit further, later on we'll provide a pluggable configuration backend. I'd rather not rely on the on-disk .git/config file to hold the params for another configuration backend.

In the end, I *think* storing/retrieving the required params to initialize a backend is a concern which should be seen as orthogonal to the backend itself or its initialization (if I rent a safe in a bank, I am responsible for the management of the key, not the bank, nor the safe. I *could* store the key in another safe next to the first one, but eventually, no matter how many safes I rent, I'd still have one key to deal with...).

Long story short, I'm not sure promoting to use the config as a storage for backend parameters is such a good idea.

Besides, I'd really prefer to see all the backends passed to `RepositoryOptions` and avoid coping with potential side effects of having backends (un)plugged to the `Repository` instance mid-flight.

**Comment by nulltoken** [comment]
@ethomson Sorry for the wall of text :/

**Comment by ethomson** [comment]
There's two issues here:  "In memory" repositories and repositories that have custom object databases.  While an "in memory" repository *must* have a custom object database, there is nothing stopping a "regular" repository from having one.  You could have a database as an *additional* source of objects besides your loose/pack sources, for example, which might be useful for keeping the size of your on-disk repository down and additional sources could be retrieved lazily.

While yes, a purely in memory repository can probably take its registration at initialization without too much difficulty, we should not restrict from the ability to add a new backend to a "normal" repository, which is something that you can do now quite easily.

**Comment by nulltoken** [comment]
>While yes, a purely in memory repository can probably take its registration at initialization without too much difficulty, we should not restrict from the ability to add a new backend to a "normal" repository, which is something that you can do now quite easily.

I agree.

A "normal" repository may be seen as opened through an optional `RepositoryOptions`. The default `RepositoryOptions` would be initialized with some `FileSystemOdbBackend`, `FileSystemConfigBackend`, `FileSystemRefBackend` with a priority of 1.

Adding a new backend to a normal repository would be a matter of adding a new backend instance to the list with a higher priority.

Building a pure "custom" repository would require to clear the list from the `RepositoryOptions` instance and add the new backends.

**Comment by ethomson** [comment]
Okay, well...  I don't understand your opposition to my desire to configure my odb backend in my configuration.  The repository configuration seems the absolutely most logical place to put it.  Can you help me understand?

**Comment by nulltoken** [comment]
>I don't understand your opposition to my desire to configure my odb backend in my configuration.

I wouldn't dare oppose to you desire :wink:

However, I'd prefer to have the backends initialized before the `Repository` is instantiated and not allow a backend to be plugged in a live `Repository`.

If one decides to store the params in the configuration, then reading those from the configuration should be done prior to initializing the backends.

This could be be done by opening a `Repository`, reading the params from the on-disk config store, disposing the repository, setting up the backend and instantiating a new `Repository` with it.

Another option may be to extend the public `Configuration` constructor and make it also accept a optional local config file path (ie. `.git/config`).

**Comment by ethomson** [comment]
> However, I'd prefer to have the backends initialized before the Repository is instantiated and not allow a backend to be plugged in a live Repository.

We rely *heavily* on the existing API that has been extant for *years*.  Breaking this API would be *terrible* for us.

**Comment by nulltoken** [comment]
>>However, I'd prefer to have the backends initialized before the Repository is instantiated and not allow a backend to be plugged in a live Repository.

>We rely *heavily* on the existing API that has been extant for *years*.  Breaking this API would be *terrible* for us.

We could deprecate the setter in v1.0 and drop it in v2.0 (With a proposed period of one year between each major version).

**Comment by jamill** [comment]
> However, I'd prefer to have the backends initialized before the Repository is instantiated and not allow a backend to be plugged in a live Repository.

What about a callback that can be passed in that gets called during repository initialization to allow consumers to setup backends. I think this would allow for everything to be initialized before the repository is "used", but it would allow consumer to use the configs the repository uses.

**Comment by ethomson** [comment]
So what's the process forward here?  Are we going to merge this soon(ish) so that we can mark it as deprecated?  Or is your desire to wait until we have wrapped all the other stuff before we can merge this?

File: LibGit2Sharp/Repository.cs [hunk 2]
```diff
@@ -37,28 +37,93 @@ public sealed class Repository : IRepository
...
             try
             {
-                handle = Proxy.git_repository_open(path);
+                handle = (path != null) ? Proxy.git_repository_open(path) : Proxy.git_repository_new();
                 RegisterForCleanup(handle);
 
                 isBare = Proxy.git_repository_is_bare(handle);
 
+                /* TODO: bug in libgit2, update when fixed by
+                 * https://github.com/libgit2/libgit2/pull/2970
```
Comment (nulltoken) [inline LibGit2Sharp/Repository.cs]: @ethomson As this is merged in the current version of libgit2 we're working with, should we remove this workaround?

**Comment by nulltoken** [comment]
@carlosmn Earlier you mentioned something about some potential issue when writing against those in-memory repos. Could you please elaborate on this?

**Comment by carlosmn** [comment]
It comes from this [SO question](https://stackoverflow.com/questions/28978460/libgit2-write-tree-object-to-mysql-odb-backend) where we assume we have a path for the repository instance, but if you create an instance yourself, that path is going to be `NULL`.

**Comment by carlosmn** [comment]
I've taken the liberty of merging mater into this branch to ask CI whether the test still pass everywhere. I'll merge manually once it reports back green.

**Comment by carlosmn** [comment]
Merged manually into master.
